File: cosmos/util/message.go
package util

import (
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"

	"github.com/dan13ram/wpokt-oracle/app"
	"github.com/dan13ram/wpokt-oracle/common"
	"github.com/dan13ram/wpokt-oracle/models"
)

func NewMessageBody(
	senderAddress string,
	amount uint64,
	recipientAddress string,
) models.MessageBody {
	return models.MessageBody{
		SenderAddress:    senderAddress,
		Amount:           amount,
		RecipientAddress: recipientAddress,
	}
}

func NewMessageContent(
	version uint8,
	nonce uint32,
	originDomain uint32,
	sender string,
	destinationDomain uint32,
	recipient string,
	messageBody models.MessageBody,
) models.MessageContent {
	return models.MessageContent{
		Version:           version,
		Nonce:             nonce,
		OriginDomain:      originDomain,
		Sender:            sender,
		DestinationDomain: destinationDomain,
		Recipient:         recipient,
		MessageBody:       messageBody,
	}
}

func NewMessage(
	originTransaction *primitive.ObjectID,
	originTransactionHash string,
	messageID string,
	content models.MessageContent,
	signatures []models.Signature,
	transaction primitive.ObjectID,
	sequence uint64,
	status models.MessageStatus,
	transactionHash string,
) *models.Message {
	return &models.Message{
		OriginTransaction:     originTransaction,
		OriginTransactionHash: originTransactionHash,
		MessageID:             messageID,
		Content:               content,
		Signatures:            signatures,
		Transaction:           transaction,
		Sequence:              sequence,
		Status:                status,
		TransactionHash:       transactionHash,
		CreatedAt:             time.Now(),
		UpdatedAt:             time.Now(),
	}
}

func UpdateMessage(messageID *primitive.ObjectID, update bson.M) error {
	if messageID == nil {
		return fmt.Errorf("messageID is nil")
	}
	return app.DB.UpdateOne(
		common.CollectionMessages,
		bson.M{"_id": messageID},
		bson.M{"$set": update},
	)
}

File: e2e/tester/src/types/message.ts
// Import the required dependencies for primitive types and date handling
import { ObjectId } from 'mongodb';
import { Hex } from 'viem';

// Define the MessageContent type
export type MessageContent = {
  version: number;
  nonce: number;
  origin_domain: number;
  sender: Hex;
  destination_domain: number;
  recipient: Hex;
  message_body: MessageBody;
};

// Define the MessageStatus type
export type MessageStatus = 'pending' | 'signed' | 'broadcasted' | 'success' | 'invalid';

// Define the Message type
export type Message = {
  id?: ObjectId;
  origin_transaction: ObjectId;
  origin_transaction_hash: Hex;
  messageId: Hex;
  content: MessageContent;
  transactionBody: Hex;
  signatures: Signature[];
  transaction?: ObjectId;
  sequence?: number;
  transaction_hash: Hex;
  status: MessageStatus;
  created_at: Date;
  updated_at: Date;
};

// Define the MessageBody type
export type MessageBody = {
  sender_address: Hex;
  amount: number;
  recipient_address: Hex;
};

// Define the Signature type
export type Signature = {
  signer: Hex;
  signature: Hex; // Assuming signature is a string representation
};

export const CollectionMessages = 'messages';
